### ISSUE TITLE
Token renewal fails while/after debugging

### ISSUE THREAD

**Issue by Vity01** [spring-projects/spring-vault#530]
Hello @mp911de 
My current status:
- Hashicorp Vault
- Spring Vault 2.1.3
- authentication: APPROLE with secretId,roleId
- token expires in 60 seconds
- I don't have any special beans instantiated, just default ones from the autoconfiguration

My problem is similar to this issue: https://github.com/spring-projects/spring-vault/issues/222
Everything is working fine (I call issueCertificate endpoint) until I put a breakpoint into the code (no matter which location - it just stops the app) - after this renewal token interruption all subsequent issueCertificate calls ends with 403 permission denied. I expect to get these 403s only for the next minute, but it looks like the token is never renewed back to the valid state.

Can you help me and point me out how to investigate this problem? Why the token is not renewed again?
For me it looks like a serious bug in correct behavior.

**Comment by Vity01** [comment]
I have a few more observations from the problem debugging. 
It looks like the server is closing connection immediately once the token is outdated and SVault is not able to recover from it. The token is neither renewed (no other attempt) again nor got new one from the re-login. 
```
2020-01-22 09:03:30.275 DEBUG 17588 --- [ctor-http-nio-4] v.a.ReactiveLifecycleAwareSessionManager : Cannot renew token, resetting token and performing re-login on next token access: reactor.netty.http.client.PrematureCloseException: Connection prematurely closed BEFORE response
2020-01-22 09:03:30.285 ERROR 17588 --- [ctor-http-nio-4] v.a.ReactiveLifecycleAwareSessionManager : Cannot renew VaultToken

org.springframework.vault.authentication.VaultTokenRenewalException: Cannot renew token; nested exception is reactor.netty.http.client.PrematureCloseException: Connection prematurely closed BEFORE response
	at org.springframework.vault.authentication.ReactiveLifecycleAwareSessionManager.lambda$renewToken$5(ReactiveLifecycleAwareSessionManager.java:242) ~[spring-vault-core-2.1.3.RELEASE.jar:2.1.3.RELEASE]
...
Caused by: reactor.netty.http.client.PrematureCloseException: Connection prematurely closed BEFORE response
	Suppressed: reactor.core.publisher.FluxOnAssembly$OnAssemblyException: 
Error has been observed at the following site(s):
	|_ checkpoint ⇢ Request to POST https://vault...:443/v1/auth/token/renew-self [DefaultWebClient]
```

**Comment by Vity01** [comment]
Any idea what's causing this token caching? It's not easy to debug, since the class structure is complex and it's mixing reactive and classical approach...

**Comment by mp911de** [comment]
What type of setup are you using? Spring Cloud Vault, Spring Vault with `AbstractReactiveVaultConfiguration`?
Typically, in a mixed setup, `VaultTemplate.sessionManager` defaults to `ReactiveSessionManagerAdapter` which is backed by `ReactiveSessionManager`.

`ReactiveLifecycleAwareSessionManager.getVaultToken()` is called by `ReactiveVaultTemplate` via the session filter that gets invoked on each Vault request that is issued through `ReactiveVaultTemplate`.

I cannot exactly tell why the connection gets disconnected while stopping the application in a debugger. My best guess is due to SSL handshake timeouts.

Looking through our code, especially `AbstractReactiveVaultConfiguration`, I noticed that `vaultTokenSupplier()` returns a [cached `VaultTokenSupplier` that wraps `ClientCertificateAuthentication`](https://github.com/spring-projects/spring-vault/blob/c41403a68981931014dfd03aaccfc4923521187f/spring-vault-core/src/main/java/org/springframework/vault/config/AbstractReactiveVaultConfiguration.java#L155). If you're using `AbstractReactiveVaultConfiguration`, it would make sense to override that method and return just a plain `AuthenticationStepsOperator` that is configured with the certificate auth method.

**Comment by Vity01** [comment]
I am sorry Mark, I inserted this bug report into the bad project. I am using `org.springframework.cloud:spring-cloud-starter-vault-config` but I see you are maintaining both.

Ad SSL handshake timeouts - that's weird - the token refresh Task is launched when I resume the app from scratch, the POST request is sent from scratch, but I don't know what's cached inside of webclient.

Ad setup - according to debugging, `VaultReactiveBootstrapConfiguration` is used and the session manager is equal to `sessionManager.getSessionToken()::block` where the session manager is wrapped by ReactiveSessionManager.  
I suspected `CachingVaultTokenSupplier` as well, but its `getVaultToken()` is not called after the fatal exception. At this moment I suspect something like invalid usage Reactor's hot vs cold call problem, but I have no evidence for it yet.

I just joined to https://gitter.im/spring-projects/spring-boot if you want to talk more interactively.
